Sources/Commands/SwiftTestTool.swift

```diff
@@ -93,32 +135,55 @@ public class SwiftTestTool: SwiftTool<TestToolOptions> {
         case .version:
             print(Versioning.currentVersion.completeDisplayString)
 
-        case .listTests:
+        case .listTests(let filter):
             let testPath = try buildTestsIfNeeded(options)
             let testSuites = try getTestSuites(path: testPath)
+            let tests = testSuites.filteredTests(specifier: filter)
+
             // Print the tests.
-            for testSuite in testSuites {
-                for testCase in testSuite.tests {
-                    for test in testCase.tests {
-                        print(testCase.name + "/" + test)
-                    }
-                }
+            for test in tests {
+                print(test.specifier)
             }
 
-        case .run(let specifier):
-            let testPath = try buildTestsIfNeeded(options)
-            let success: Bool = TestRunner(path: testPath, xctestArg: specifier, processSet: processSet).test()
-            exit(success ? 0 : 1)
+        case .runSpecific(let specifier):
+            // This is a deprecated option, so emit diagnostics accordingly
+            diagnostics.emit(data: SpecifierDeprecatedDiagnostic())
+            let ranSuccessfully = try runTestsSerially(specifier: specifier)
+            exit(ranSuccessfully ? 0 : 1)
+
+        case .runSerial(let specifier):
+            let ranSuccessfully = try runTestsSerially(specifier: specifier)
+            exit(ranSuccessfully ? 0 : 1)
 
-        case .runInParallel:
+        case .runParallel(let filterString):
             let testPath = try buildTestsIfNeeded(options)
             let testSuites = try getTestSuites(path: testPath)
+            let tests = testSuites.filteredTests(specifier: filterString)
             let runner = ParallelTestRunner(testPath: testPath, processSet: processSet)
-            try runner.run(testSuites)
+            try runner.run(tests)
             exit(runner.ranSuccesfully ? 0 : 1)
         }
     }
 
+    /// Executes the tests, optinally filtered by specifier, serially
+    /// - Returns: true if success, otherwise false
+    private func runTestsSerially(specifier: TestCaseSpecifier) throws -> Bool {
+        let testPath = try buildTestsIfNeeded(options)
+        let testSuites = try getTestSuites(path: testPath)
+        let tests = testSuites.filteredTests(specifier: specifier)
+
+        var isSuccess = true
+        for test in tests {
+            let runner = TestRunner(path: testPath,
```

Comment:
We should create a fast path if specifier is `none` i.e. avoid launching a new subprocess for each test case unless really needed.